File: src/app/services/field-intelligence.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  DashboardSummary,
  Field,
  FieldsResponse,
  NutrientBalance,
  YieldPrediction,
  OperationsResponse,
  SoilTestsResponse,
  GeoJsonFeatureCollection,
} from '../models/field-intelligence.models';

const BASE = '/field-api';

@Injectable({ providedIn: 'root' })
export class FieldIntelligenceService {
  private http = inject(HttpClient);

  getDashboardSummary(): Observable<DashboardSummary> {
    return this.http.get<DashboardSummary>(`${BASE}/api/dashboard/summary`);
  }

  getFields(): Observable<FieldsResponse> {
    return this.http.get<FieldsResponse>(`${BASE}/api/fields`);
  }

  getField(fieldId: string): Observable<{ field: Field }> {
    return this.http.get<{ field: Field }>(`${BASE}/api/fields/${fieldId}`);
  }

  getNutrientBalance(fieldId: string): Observable<NutrientBalance> {
    return this.http.get<NutrientBalance>(`${BASE}/api/fields/${fieldId}/nutrient-balance`);
  }

  getYieldPrediction(fieldId: string): Observable<YieldPrediction> {
    return this.http.get<YieldPrediction>(`${BASE}/api/fields/${fieldId}/yield-prediction`);
  }

  getOperations(fieldId: string): Observable<OperationsResponse> {
    return this.http.get<OperationsResponse>(`${BASE}/api/fields/${fieldId}/operations`);
  }

  getSoilTests(fieldId: string): Observable<SoilTestsResponse> {
    return this.http.get<SoilTestsResponse>(`${BASE}/api/fields/${fieldId}/soil-tests`);
  }

  getGisBlocks(): Observable<GeoJsonFeatureCollection> {
    return this.http.get<GeoJsonFeatureCollection>(`${BASE}/api/gis/blocks`);
  }
}

File: Services/FieldIntelligenceService.cs
using System.Net.Http.Json;
using AgronomyStudio.Models;

namespace AgronomyStudio.Services;

public sealed class FieldIntelligenceService
{
    private readonly HttpClient _http;

    public FieldIntelligenceService(IHttpClientFactory factory)
    {
        _http = factory.CreateClient(ApiClients.Field);
    }

    public Task<DashboardSummary?> GetDashboardSummaryAsync(CancellationToken ct = default) =>
        _http.GetFromJsonAsync<DashboardSummary>("api/dashboard/summary", ct);

    public Task<FieldsResponse?> GetFieldsAsync(CancellationToken ct = default) =>
        _http.GetFromJsonAsync<FieldsResponse>("api/fields", ct);

    public Task<NutrientBalance?> GetNutrientBalanceAsync(string fieldId, CancellationToken ct = default) =>
        _http.GetFromJsonAsync<NutrientBalance>($"api/fields/{fieldId}/nutrient-balance", ct);

    public Task<YieldPrediction?> GetYieldPredictionAsync(string fieldId, CancellationToken ct = default) =>
        _http.GetFromJsonAsync<YieldPrediction>($"api/fields/{fieldId}/yield-prediction", ct);

    public Task<OperationsResponse?> GetOperationsAsync(string fieldId, CancellationToken ct = default) =>
        _http.GetFromJsonAsync<OperationsResponse>($"api/fields/{fieldId}/operations", ct);

    public Task<SoilTestsResponse?> GetSoilTestsAsync(string fieldId, CancellationToken ct = default) =>
        _http.GetFromJsonAsync<SoilTestsResponse>($"api/fields/{fieldId}/soil-tests", ct);

    public Task<GeoJsonFeatureCollection?> GetGisBlocksAsync(CancellationToken ct = default) =>
        _http.GetFromJsonAsync<GeoJsonFeatureCollection>("api/gis/blocks", ct);
}
